sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/models/BinaryContent.java

```diff
@@ -10,7 +10,7 @@
  * An abstraction for a message containing a content type along with its data.
  */
 @Fluent
-public class MessageWithMetadata {
+public class BinaryContent {
```

Comment:
We would now have BinaryData and BinaryContent as public types and another implementation type i.e. BinaryDataContent. This is likely going to confuse us in future :)

Reply:
This was based on architect feedback. The other languages also have moved their names to this. I can add you to the thread if you want.

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs#L9

Reply:
The idea is that BinaryContent contains data along with the content type. We didn't think it would be confusing as they are two related and similarly named types, rather than being just two similarly named types.

Reply:
/cc @tg-msft

Reply:
Should `BinaryData` be able to contain content type ?

Reply:
And if not should this be just `Content` ? If `BinaryContent` contains `BinaryData` that contains `text/plain` then is it still binary or not ? :-)

Reply:
BinaryData can contain textual data or structured data - it is just represented as bytes. BinaryContent is a way to attach a content type to BinaryData.

Reply:
> Should `BinaryData` be able to contain content type ?

We considered something like this in the past, but the semantics become confusing when dealing with unknown content types/formats, e.g. should unknown always be considered binary? By having it on a separate wrapping type it can be used in an opt-in way for scenarios that will always have a content type.

Reply:
@pallavit, @srnagar, could you please elaborate what's confusing here and what concrete implications of this confusion you see? There are many types in all platforms that are similar, yet they are not considered confusing, e.g. Int32 and Int64 are super similar.

Reply:
Java already has [BinaryDataContent](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/BinaryDataContent.java) which is wrapped by `BinaryData`. Also, when deciding which type to use, would there be a scenario where we want to use `BinaryData` and not `BinaryContent` given that content type in BinaryContent is optional? If this type is intended for extension in Event Hubs and Service Bus, should we make this type abstract?

Reply:
It was initially abstract, but we made it concrete to make it easier for use with Schema Registry that didn't involve Service Bus/Event Hubs. However, I think we can still make this scenario easy by having a private implementation that we return if the user doesn't specify a more specific BinaryContent type. This would allow us to move the type back to being abstract which I believe would address your concerns around any confusion that might arise from how this type is meant to be used.

Reply:
Here is what this change would look like in .NET - https://github.com/Azure/azure-sdk-for-net/pull/27525

After making the class abstract, I needed to add two additional generic overloads that only took the TData instead of both the TEnvelope and TData. When using the new overloads we would construct the internal DefaultBinaryContent and return that to the user. This was necessary in order to leave in our generic type constraint that TEnvelope have an empty public constructor.